### PR TITLE
HttpRequestorGem: Add basic autogen documentation to headers

### PR DESCRIPTION
**Pull request by lmbr-pip**
This is just a documentation change, deliberately not changing the code for now.

Signed-off-by: rppotter <rppotter@amazon.com>

### CODE REVIEW

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestorBus.h [hunk 1]
```diff
@@ -5,32 +5,78 @@
...
+    //! See [HTTP RFC](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html) for expectations around methods, headers, and body.
+    class HttpRequestorRequests : public AZ::EBusTraits
     {
-
     public:
         static const AZ::EBusHandlerPolicy HandlerPolicy = AZ::EBusHandlerPolicy::Single;
         static const AZ::EBusAddressPolicy AddressPolicy = AZ::EBusAddressPolicy::Single;
 
-        // Public functions
+        //! Make a restful call to a HTTP(s) endpoint. Will get response as JSON.
```
[the remark this answers is not in the document]
Comment (lmbr-pip) [reply Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestorBus.h:23]: Will clarify that the callback will receive the response as JSON

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h [hunk 2]
```diff
@@ -34,50 +49,77 @@ namespace HttpRequestor
         TextParameters(TextParameters&&) = default;
         TextParameters& operator=(TextParameters&&) = default;
 
-        //returns the URI in string form as an recipient of the HTTP connection
-        const Aws::String& GetURI() const { return m_URI; }
+        //! Get the URI in string form as an recipient of the HTTP connection.
+        //! @return URI as string
```
Comment (sptramer) [inline Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h:53]:
```suggestion
        //! Get the endpoint URI as a string.
```

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h [hunk 3]
```diff
@@ -34,50 +49,77 @@ namespace HttpRequestor
...
+        const Aws::String& GetURI() const
+        {
+            return m_URI;
+        }
 
-        //returns the method of which the HTTP request will take. GET, POST, DELETE, PUT, or HEAD
-        Aws::Http::HttpMethod GetMethod() const { return m_method; }
+        //! Get the method of which the HTTP request will take. GET, POST, DELETE, PUT, or HEAD.
+        //! @return The http method to use
```
Comment (sptramer) [inline Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h:60]:
```suggestion
        //! Get the HTTP method used by the request.
```

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h [hunk 4]
```diff
@@ -34,50 +49,77 @@ namespace HttpRequestor
...
+        Aws::Http::HttpMethod GetMethod() const
+        {
+            return m_method;
+        }
 
-        //returns the list of extra headers to include in the request
-        const Headers & GetHeaders() const { return m_headers; }
+        //! Get the list of extra headers to include in the request
+        //! @return map of header name and values to use.
```
Comment (sptramer) [inline Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h:67]:
```suggestion
        //! Get the list of extra headers to include in the request
        //! @return Map of header-value pairs.
```

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h [hunk 5]
```diff
@@ -34,50 +49,77 @@ namespace HttpRequestor
...
+        const Headers& GetHeaders() const
+        {
+            return m_headers;
+        }
 
-        //returns the stream for the body of the request
-        const std::shared_ptr<std::stringstream> & GetBodyStream() const { return m_bodyStream; }
+        //! Get the stream for the body of the request
+        //! @return a content body stream that can be used to send the body to the endpoint.
```
Comment (sptramer) [inline Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h:74]:
```suggestion
        //! Get the input stream used to send the body of the request to the endpoint.
        //! @return A string stream representing the request body.
```

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h [hunk 6]
```diff
@@ -34,50 +49,77 @@ namespace HttpRequestor
...
+        const std::shared_ptr<std::stringstream>& GetBodyStream() const
+        {
+            return m_bodyStream;
+        }
 
-        //returns the function of which to feed back the TEXT that the HTTP call resulted in. The function also requires the HTTPResponseCode indicating if the call was successful or failed
-        const TextCallback & GetCallback() const { return m_callback; }
+        //! Get the callback function of which to feed back the TEXT that the HTTP call resulted in.
+        //! The function also requires the HTTPResponseCode indicating if the call was successful or failed.
+        //! @return the callback function to use to return the response from the endpoint.
```
Comment (sptramer) [inline Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h:82]:
```suggestion
        //! Get the callback function for processing the text returned in an HTTP response.
        //! Callback functions are responsible for correctly interpreting the HTTP response code, and should communicate any failures.
//! @return The callback function to process endpoint responses with.
```

**Review by sptramer — CHANGES_REQUESTED** [review]
Some wording changes and expansions throughout, and a couple questions about how the `body` stream is read during HTTP connection/transmission.

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestParameters.h [hunk 7]
```diff
@@ -36,30 +50,50 @@ namespace HttpRequestor
...
     private:
-        Aws::String                             m_URI;
-        Aws::Http::HttpMethod                   m_method;
-        Headers                                 m_headers;
-        std::shared_ptr<std::stringstream>      m_bodyStream;   // required by Aws::Http::HttpRequest
-        Callback                                m_callback;
+        Aws::String m_URI;
+        Aws::Http::HttpMethod m_method;
+        Headers m_headers;
+        std::shared_ptr<std::stringstream> m_bodyStream; // required by Aws::Http::HttpRequest
```
[the remark this answers is not in the document]
Comment (lmbr-pip) [reply Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestParameters.h:93]: Agreed, we should only provide public documentation in the API reference.

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestParameters.h [hunk 8]
```diff
@@ -12,21 +12,35 @@
...
+        //! @param method The http verb to use, for example HTTP_GET.
+        //! @param headers A map of header names and values to use.
+        //! @param callback The callback method to receive a HTTP call's response.
         Parameters(const AZStd::string& URI, Aws::Http::HttpMethod method, const Headers& headers, const Callback& callback);
-        Parameters(const AZStd::string& URI, Aws::Http::HttpMethod method, const Headers& headers, const AZStd::string& body, const Callback& callback);
+
+        //! @param URI A universal resource indicator representing an endpoint.
+        //! @param method The http verb to use, for example HTTP_POST.
+        //! @param headers A map of header names and values to use.
+        //! @param body An data to associate with an HTTP call.
```
[the remark this answers is not in the document]
Comment (lmbr-pip) [reply Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestParameters.h:36]: You can technically send a body on everything other than TRACE and HEAD I believe: https://stackoverflow.com/questions/16339198/which-http-methods-require-a-body

Most of this phrasing is from the AWS C++ SDK headers as we're mostly passing through to the underlying code.

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestParameters.h [hunk 9]
```diff
@@ -36,30 +50,50 @@ namespace HttpRequestor
...
+        }
+
+        //! Get the method of which the HTTP request will take. GET, POST, DELETE, PUT, or HEAD.
+        //! @return The http method to use
+        Aws::Http::HttpMethod GetMethod() const
+        {
+            return m_method;
+        }
+
+        //! Get the list of extra headers to include in the request
```
[the remark this answers is not in the document]
Comment (lmbr-pip) [reply Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestParameters.h:67]: Fixed.

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestParameters.h [hunk 10]
```diff
@@ -36,30 +50,50 @@ namespace HttpRequestor
...
+        //! @return map of header name and values to use.
+        const Headers& GetHeaders() const
+        {
+            return m_headers;
+        }
+
+        //! Get the stream for the body of the request
+        //! @return a content body stream that can be used to send the body to the endpoint.
```
[the remark this answers is not in the document]
Comment (lmbr-pip) [reply Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestParameters.h:75]: Thats an implementation detail of the underlying AWS C++ SDK, https://sdk.amazonaws.com/cpp/api/LATEST/class_aws_1_1_http_1_1_http_request.html#a4bd6b611026e33dc697d716cec7403cd  - we just need to provide it and it needs to be valid for the length of the request.

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestorBus.h [hunk 11]
```diff
@@ -5,32 +5,78 @@
...
+        //! @param URI The universal resource indicator representing the endpoint to make the request to.
+        //! @param method The http verb to use, for example HTTP_GET.
+        //! @param callback The callback method to receive the JSON response object.
```
[the remark this answers is not in the document]
Comment (lmbr-pip) [reply Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestorBus.h:26]: Fixed RESTful. Note: Client can request content type returned, not sure we state endpoint is required to provide format requested.

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestParameters.h [hunk 12]
```diff
@@ -36,30 +50,49 @@ namespace HttpRequestor
...
-
-        //returns the list of extra headers to include in the request
-        const Headers & GetHeaders() const { return m_headers; }
-
-        //returns the stream for the body of the request
-        const std::shared_ptr<std::stringstream> & GetBodyStream() const { return m_bodyStream; }
-
-        //returns the function of which to feed back the JSON that the HTTP call resulted in. The function also requires the HTTPResponseCode indicating if the call was successful or failed
-        const Callback & GetCallback() const { return m_callback; }
+        //! Get the URI in string form as an recipient of the HTTP connection.
```
Comment (junbo75) [inline Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestParameters.h:53]: Nit: Do we want to add comment for the return value?

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestParameters.h [hunk 13]
```diff
@@ -36,30 +50,49 @@ namespace HttpRequestor
...
+        const Aws::String& GetURI() const
+        {
+            return m_URI;
+        }
+
+        //! Get the HTTP method configured to use for a request.
+        Aws::Http::HttpMethod GetMethod() const
```
Comment (junbo75) [inline Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestParameters.h:60]: Nit: Same as above. Do we want to add comment for the return value?

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h [hunk 14]
```diff
@@ -34,50 +49,76 @@ namespace HttpRequestor
         TextParameters(TextParameters&&) = default;
         TextParameters& operator=(TextParameters&&) = default;
 
-        //returns the URI in string form as an recipient of the HTTP connection
-        const Aws::String& GetURI() const { return m_URI; }
+        //! Get the URI in string form as an recipient of the HTTP connection.
```
Comment (junbo75) [inline Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h:52]: Nit: Same here. Comment for the return value?

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h [hunk 15]
```diff
@@ -34,50 +49,76 @@ namespace HttpRequestor
...
+        const Aws::String& GetURI() const
+        {
+            return m_URI;
+        }
 
-        //returns the method of which the HTTP request will take. GET, POST, DELETE, PUT, or HEAD
-        Aws::Http::HttpMethod GetMethod() const { return m_method; }
+        //! Get the HTTP method configured to use for a request.
+        Aws::Http::HttpMethod GetMethod() const
```
Comment (junbo75) [inline Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h:59]: Nit: Same here. Comment for the return value?

**Review by junbo75 — APPROVED** [review]

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h (hunk 14 above)
Comment (lmbr-pip) [reply Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h:52]: @sptramer asked me to remove the return value documentation as it added nothing for simple types.

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestParameters.h (hunk 13 above)
Comment (lmbr-pip) [reply Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestParameters.h:60]: @sptramer asked me to remove the return value documentation as it added nothing for simple types.

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h (hunk 15 above)
Comment (lmbr-pip) [reply Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTextRequestParameters.h:59]: @sptramer asked me to remove the return value documentation as it added nothing for simple types.

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestParameters.h (hunk 12 above)
Comment (lmbr-pip) [reply Gems/HttpRequestor/Code/Include/HttpRequestor/HttpRequestParameters.h:53]: @sptramer asked me to remove the return value documentation as it added nothing for simple types.

**Review by onecent1101 — APPROVED** [review]

File: Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTypes.h [hunk 16]
```diff
@@ -23,20 +23,13 @@ AZ_POP_DISABLE_WARNING
 
 namespace HttpRequestor
 {
-    //
-    // the call back function for http requests.
-    //
+    // The call back function for http requests.
```
[the remark this answers is not in the document]
Comment (lmbr-pip) [reply Gems/HttpRequestor/Code/Include/HttpRequestor/HttpTypes.h]: Fixed

**Review by sptramer — APPROVED** [review]

**Pull request merged**